### ISSUE TITLE
Debug mode for ignite

### ISSUE THREAD

**Issue by sdesrozis** [pytorch/ignite#1992]
## 🚀 Feature

The idea is to provide a debug mode as discussed here: https://github.com/pytorch/ignite/issues/1989#issuecomment-833643158.

All the possible options can be discussed here. We could add an helper to track the workflow of handlers from the engine run. On the other hand, debug tools are needed to track the values of PyTorch objects (LR, loss, grads, etc.). Note that the preferred way to extend ignite is handlers' design so let's create a new set of tools for the debugging.

**Comment by vfdev-5** [comment]
We could also think of different levels of debugging:

- LR, loss and grads will be essential for training low-level debugging.
- Dataflow debugging : 
  - printing input shapes, dtypes, devices, data stats
  - saving input data as tensors or images for vision, etc
- usage of pytorch built-in tools
- ...

**Comment by Ishan-Kumar2** [comment]
Hi @sdesrozis and @vfdev-5, I think this is a great idea and much needed tool. I am interested in implementing it.

I agree with @vfdev-5's idea of having it in levels. We can pass the level in `engine.run`. So level 1 could be dataflow/workflow debugging, that means running the engine and logging all the events and handlers.
Something like 
```EPOCH STARTED
ITERATION STARTED
   Input Shape - 
   Handler 1 called
   ....
ITERATION COMPLETED
   Handler 2 called
...
```
Doing this is start forward, we just need to set the logger level in `engine.run`

The level 2 would be logging information during training like loss, LR, scheduler. For this we need to make changes to the `training_step` the user provides, I am not sure how we can go about that.

Then level 3 would be gradients. (can use PyTorch hooks) 

What do you think?

**Comment by DevPranjal** [comment]
Hello, @sdesrozis @vfdev-5. Is this feature request still on the cards? I would love to discuss further and contribute.

**Comment by vfdev-5** [comment]
@DevPranjal yes, it is still open issue that we would like to have.

**Comment by DevPranjal** [comment]
Great, so if I can compile the previous discussions, we need level wise debugging modes:
- [ ] **Level 1**: simple event and handler prints, output saving and visualization
- [ ] **Level 2**: tracking loss, lr, scheduler calls
- [ ] **Level 3**: tracking gradients, updates

@vfdev-5 The debug level API will be accessed through `engine.run(...)`?

**Comment by vfdev-5** [comment]
I was thinking about `engine.debug(enabled=True, level=level)` and `engine.debug(enabled=False)`.

**Comment by vfdev-5** [comment]
Yes, you are right maybe a binary flag would be better option:
```python
# to enable
engine.debug(level=Engine.DEBUG_EVENTS | Engine.DEBUG_OUTPUT | Engine.DEBUG_GRADS)

# to disable
engine.debug(level=0)
# or engine.debug(level=None)
```

Provided flag names are just examples and can be renamed if a better naming can be found

**Comment by DevPranjal** [comment]
Sure.
For a start, I tried the following:
```python
def debug(self, level: int = 0):
    self.debug_level = level

def _print_debug_mode(self, string: str, level: int):
    if level > self.debug_level:
        print(string)

def _fire_event(...):
    ...
    self._print_debug_mode(f'firing handler: {func.__name__}', level=1)
    ...
```

**Comment by vfdev-5** [comment]
@DevPranjal there is also `self.logger` https://github.com/pytorch/ignite/blob/3a286b1d13a4a0476b3ee8e8046f16465818c9f6/ignite/engine/engine.py#L125

**Comment by DevPranjal** [comment]
Yes.

Can I change the approach to have custom log levels, using `logging.addLevelName` for all the three levels?

or do I simply change the above to:
```python
def _log_debug_mode(self, string: str, level: int):
    if level > self.debug_level:
        self.logger.debug(string)
```
or anything else?

**Comment by sawradip** [comment]
Hello @puhuk, are you  solving this Logging issue? Or is it open?

**Comment by sawradip** [comment]
Does @puhuk 's PR solve this issue, or is this open yet?

**Comment by puhuk** [comment]
Hi, @sawradip I’m working on it :)

**Comment by vfdev-5** [comment]
@puhuk @sawradip the fact that someone is working on this issue does not mean that produced PR will be certainly merged. It can happen that contributor stops working on the PR and PR is suspended or PR's code does not meet requirement or has poor quality...
Anyway, @sawradip you can also try to propose your version of solution. Otherwise pick another issue from the list.
